Nprocs 3 perf

jayshrivastava opened the pull request.

Commit:
span: make span.Frontier thread safe

Make `span.Frontier` thread safe by default.

Release note: None

Commit:
changefeedccl: parallelize event consumption

Previously, KV events were consumed and processed by changefeed
aggregator using a single, synchronous Go routine. This PR makes
it possible to run up to `changefeed.event_consumer_workers`
consumers to consume events concurrently. The cluster setting
`changefeed.event_consumer_worker_queue_size` is added to help
control the number of concurrent events we can have in flight.

Specifying `changefeed.event_consumer_workers=0` keeps existing
single threaded implementation.

Release note (enterprise change): This change adds the cluster setting
`changefeed.event_consumer_workers` which allows changefeeds to
process events concurrently.

jayshrivastava committed:
do not merge: force parallel event consumer for testing

This change forces 32 parallel event consumer to be used
by changefeeds for testing purposes. Once CI passes, showing
that existing functionality is maintained by parallel consumers,
this commit can be dropped, reverting the number of consumers
to be a metamorphic constant.

Release note: None

Commit:
tree: Improve performance of tree.AsJSON

Improve performance of `tree.AsJSON` method.

These improvements are important for any query that produces
large number of JSON objects, as well as to changefeeds, which
rely on this function when producing JSON encoded feed.

Most of the changes revolved around modifying underlying types
(s.a. date/timestamp types, box2d, etc) to favor using functions
that append to bytes buffer, instead of relying on slower
functions, such as `fmt.Sprintf`.  The conversion
performance improved around 5-10% for most of the types, and
as high as 50% for time types:

```
Benchmark            old t/op      new t/op    delta
AsJSON/box2d-10    578ns ± 3%    414ns ± 2%   -28.49%  (p=0.000 n=10+9)
AsJSON/box2d[]-10  1.64µs ± 3%   1.19µs ± 4%  -27.14%  (p=0.000 n=10+10)
AsJSON/time-10     232ns ± 2%    103ns ± 1%   -55.61%  (p=0.000 n=10+10)
AsJSON/time[]-10   687ns ± 4%    342ns ± 4%   -50.17%  (p=0.000 n=10+10)
```

Note: Some types in the local benchmark show slight slow down in speed.
No changes were made in those types, and in general, the encoding speed
of these types might be too fast to reliable detect changes:
```
Benchmark            old t/op      new t/op       delta
AsJSON/bool[]-10    65.9ns ± 1%   67.7ns ± 2%    +2.79%  (p=0.001 n=8+9)
```

The emphasis was also placed on reducing allocations.
By relying more heavily on a pooled FmtCtx, which contains
bytes buffer, some conversions resulted in amortized
elimination of allocations (time):
```
Benchmark               old B/op      new t/op    delta
AsJSON/timestamp-10    42.1B ± 3%      0.0B      -100.00%  (p=0.000 n=10+10)
AsJSON/timestamp[]-10  174B ± 4%      60B ± 1%   -65.75%  (p=0.000 n=10+10)
```

Release Note: None
Release Justification: performance improvement

jayshrivastava committed:
changefeedccl: update avro encoder to optionally use shared cache

This change updates the avro encodeer to optionally use a user-specified cache.
This is useful for when we have multiple encoders in parallel and want them
to share the same cache for performance reasons.

Release note: None

jayshrivastava committed:
changefeedccl: use crc to distribute work in parallel consumer

This change updates the parallel event consumer to use crc to distribute work
among workers. This change is 64% faster:

name              old time/op  new time/op  delta
ShardingByKey-10  78.1ns ± 3%  28.0ns ± 1%  -64.09%  (p=0.000 n=7+10)

It also adds a test for the sharding function.

Release note: None

jayshrivastava committed:
changefeedccl: add metrics to parallel event consumer

This change adds metrics for nanos spend flushing, nanos spent waiting
for events to be passed to the workers, and the number of buffered
events.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88281)
<!-- Reviewable:end -->

Commit:
changefeedccl: Do not block on file size based flushes

Prior to this change, cloud storage sink trigger
file sized based flush whenever new row would
would push the file size beyond configured threshold.

This had the effect of singificantly reducing the throughput
whenever such event occured -- no additional events could
be added to cloud storage sink, while the previus flush was
active.

This is not necessary.  Cloud storage sink can trigger
file based flushes asynchronously.  The only requirement
is that if a real, non file based, flush arrives, or if we
need to emit resolved timestamps, then we must wait for
all of the active flush requests to complete.

In addition, because every event added to cloud sink has
associate allocation, which is released when file is written
out, performing flushes asynchronously is safe with respect
to memory usage and accounting.

Release note (enterprise change): Changefeeds, using cloud
storage sink, now have better throughput.
Release justification: performance fix
